### PR TITLE
Fixed #27504 -- Allowed using the ORM after an error and rollback.

### PR DESCRIPTION
**Pull request by marky1991**

### CODE REVIEW

**Comment by charettes** [comment]
Buildbot, add to whitelist.

**Comment by aaugustin** [comment]
This looks all right.

We may want to leave the ticket open anyway to keep track of the same problem with nested transactions?

**Comment by marky1991** [comment]
If this is fine, is backporting to 1.10 acceptable? For now, I've had to monkeypatch transaction.rollback in my code to achieve the same, which is certainly nonideal.

File: django/db/backends/base/base.py [hunk 1]
```diff
@@ -267,6 +267,8 @@ def rollback(self):
         self._rollback()
         # A successful rollback means that the database connection works.
         self.errors_occurred = False
+        # We just rolled back, so a rollback is no longer needed
```
Comment (timgraham) [inline django/db/backends/base/base.py]: I try to phrase comments without "we", e.g. "After a rollback, a rollback is no longer needed." although it seems like maybe a comment doesn't provide much benefit?

File: tests/transactions/tests.py [hunk 2]
```diff
@@ -443,8 +443,25 @@ def test_atomic_does_not_leak_savepoints_on_failure(self):
...
+
+@skipIf(
+    connection.features.autocommits_when_autocommit_is_off,
+    "This test requires a non-autocommit mode that doesn't autocommit."
+)
+class NonAutocommitTests(TransactionTestCase):
+
+    available_apps = []
+
+    def test_orm_query_after_error(self):
```
Comment (timgraham) [inline tests/transactions/tests.py]: add "and_rollback" to test name?

File: tests/transactions/tests.py [hunk 3]
```diff
@@ -443,8 +443,25 @@ def test_atomic_does_not_leak_savepoints_on_failure(self):
...
+        """#27504 -- Cannot Make ORM Queries After an Error and Rollback In Non_autocommit Mode"""
```
Comment (timgraham) [inline tests/transactions/tests.py]: State the expected behavior in test docstrings and wrap docstrings at 79 chars per [Python style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style):
```
"""
ORM queries are allowed after an error and a rollback in non-autocommit
mode (#27504).
"""
```

File: tests/transactions/tests.py (hunk 2 above)
Comment (timgraham) [inline tests/transactions/tests.py]: I didn't see any indication this is a regression. Please add details about where the regression was introduced to the ticket.

Comment (aaugustin) [inline tests/transactions/tests.py]: As far as I know, non-automcommit mode hasn't been usable since 1.6. The docs say something that boils down "you're on your own and you aren't going to have fun" because no one could offer a use case for running a web app without autocommit so I didn't put effort into supporting that.

If there's an additional regression from 1.9 to 1.10 (unclear to me), the fix should be backported.

If not, I don't think it should be backported.

Comment (marky1991) [inline tests/transactions/tests.py]: I just ran my test in 1.6, 1.7, 1.8, and 1.9. This regression was first added in 1.9 (https://bitbucket.org/marky1991/django-test/raw/59c9ff89e4b12b4a831c36171139cb022735201b/test1.py first fails in 1.9). We didn't notice then because we are updating from 1.7 to 1.10. Non-autocommit mode has definitely worked fine up until 1.7 (we use it everywhere in our code and we use 1.7 in production).

In any case, we have an internal monkeypatch workaround and can wait to remove that until 1.11 . I will work on reponding to the review so we can close this ticket.

Comment (aaugustin) [inline tests/transactions/tests.py]: Oh. Interesting :-|

Bisecting the regression on Django's master branch with your test will show where the regression happened.

Depending on what this reveals, a backport could be in order, even if the regression is old.

Comment (timgraham) [inline tests/transactions/tests.py]: Bisected to 388bb5bd9aa3cd43825cd8a3632a57d8204f875f. I didn't finish investigating to understand why that's relevant here.

Comment (aaugustin) [inline tests/transactions/tests.py]: This is very weird. That commit seems unrelated?

Comment (timgraham) [inline tests/transactions/tests.py]: The deliberate error in the reproduction script (`u.is_active = False, # assigning a tuple to a boolean field`) doesn't raise a `ValidationError` until that commit.

Comment (marky1991) [inline tests/transactions/tests.py]: Oops, sorry, I misreported that. Indeed, ValidationError isn't raised until later, but testing with IntegrityError (similar to the test added in tests.py), works in 1.8 but fails in 1.9. Working on investigating with git bisect.
